### PR TITLE
Improve S3984 performance

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
In our [April perf evaluation](https://docs.google.com/spreadsheets/d/1uo7kSDPkbRFyoXtzBp7Piftcpda29i9mUoj-ZTluCtQ/edit#gid=1598900315) , on `nodatime`, S3984 behaved pretty badly:

<google-sheets-html-origin>
ExceptionsShouldBeUsed |  6324 ms (4%)
-- | -- |

</google-sheets-html-origin>

Also, see [internal report.](https://discuss.sonarsource.com/t/performance-problem-reported-for-csharp-post-8-9-upgrade-rules-impacted-s1128-s3984-and-s1450/7867/3). 

It's pretty obvious it will behave bad because it's querying the symbol as the first operation in the rule.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsShouldBeUsed.cs [hunk 1]
```diff
@@ -35,30 +35,25 @@ public sealed class ExceptionsShouldBeUsed : SonarDiagnosticAnalyzer
...
-                if (!createdObjectType.DerivesFrom(KnownType.System_Exception))
-                {
-                    return;
-                }
-
                 var parent = objectCreation.GetFirstNonParenthesizedParent();
                 if (parent.IsKind(SyntaxKind.ExpressionStatement))
                 {
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectCreation.GetLocation()));
+                    var createdObjectType = c.SemanticModel.GetSymbolInfo(objectCreation.Type).Symbol as INamedTypeSymbol;
```
Comment (Corniel) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsShouldBeUsed.cs:50]: `&& c.SemanticModel.GetSymbolInfo(objectCreation.Type).Symbol is INamedTypeSymbol createdObjectType`

To have less nesting?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsShouldBeUsed.cs [hunk 2]
```diff
@@ -35,30 +35,23 @@ public sealed class ExceptionsShouldBeUsed : SonarDiagnosticAnalyzer
         internal const string DiagnosticId = "S3984";
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/ExceptionsShouldBeUsed.cs:35]: pedantic: can be private

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!

**Comment by Corniel** [comment]
What about:

``` c#
public void NotCompiant()
{
    SomeException.FactoryMethod(); // returns an exception
}
```

(Agreed, not part of the improvement)

**Comment by andrei-epure-sonarsource** [comment]
Hi @Corniel - if you think there's a false negative, you can open a new ticket. But generally I think exceptions are created simply by invoking the constructor.

Validation: 
- on nodatime, the time went down from 6324 ms to 58 ms
- on efcore, went down from 2755 ms to 796 ms
- on n2cms, went down from 471 ms to 33 ms

**Comment by Corniel** [comment]
> 
> Hi @Corniel - if you think there's a false negative, you can open a new ticket. But generally I think exceptions are created simply by invoking the constructor.

I know for a fact that it is not always the case, but I agree that you should not add the issue here.

>     * on nodatime, the time went down from 6324 ms to 58 ms
>     * on efcore, went down from 2755 ms to 796 ms
>     * on n2cms, went down from 471 ms to 33 ms
Nice improvements!

**Pull request merged**